compiler/optimizer/OSRDefAnalysis.cpp

```diff
@@ -699,75 +699,280 @@ TR_OSRLiveRangeAnalysis::TR_OSRLiveRangeAnalysis(TR::OptimizationManager *manage
-      return 0;
+         traceMsg(comp(), "Should not perform OSRLiveRangeAnalysis -- Not required because we are peeking\n");
+      return false;
+      }
+   else if (!comp()->supportsInduceOSR())
+      {
+      if (comp()->getOption(TR_TraceOSR))
+         traceMsg(comp(), "Should not perform OSRLiveRangeAnlysis -- OSR is not supported\n");
+      return false;
+      }
+   else if (comp()->getOSRMode() == TR::involuntaryOSR)
+      {
+      if (comp()->getOption(TR_TraceOSR))
+         traceMsg(comp(), "Should not perform OSRLiveRangeAnlysis -- involuntary OSR mode not supported\n");
+      return false;
       }
 
-   // Determine if the analysis is necessary
-   if (!canAffordAnalysis())
+   // Skip optimization if there are no OSR points
+   TR::ResolvedMethodSymbol *methodSymbol = optimizer()->getMethodSymbol();
+   if (methodSymbol->getNumOSRPoints() == 0)
       {
       if (comp()->getOption(TR_TraceOSR))
+         traceMsg(comp(), "No OSR points, skip liveness\n");
+      return false;
+      }
+
+   return true;
+   }
+
+int32_t TR_OSRLiveRangeAnalysis::perform()
+   {
+   // Check if the opt should be performed
+   if (!shouldPerformAnalysis())
+      return 0;
+
+   if (comp()->getOption(TR_TraceOSR))
+      traceMsg(comp(), "OSR reaching live range analysis can be done\n");
+
+   bool containsAuto = false, sharesParm = false, containsPendingPush = false;
+
+   // Detect autos, pending push temps and whether there is a shared parm slot
+   TR::ResolvedMethodSymbol *methodSymbol = optimizer()->getMethodSymbol();
+   TR_Array<List<TR::SymbolReference>> *autosListArray = methodSymbol->getAutoSymRefs();
+   for (uint32_t i = 0; autosListArray && i < autosListArray->size(); ++i)
+      {
+      List<TR::SymbolReference> autosList = (*autosListArray)[i];
+      ListIterator<TR::SymbolReference> autosIt(&autosList);
+      for (TR::SymbolReference *symRef = autosIt.getFirst(); symRef; symRef = autosIt.getNext())
          {
-         traceMsg(comp(), "%s OSR live range analysis is not affordable for this method because of the expected cost\n",
-            optimizer()->getMethodSymbol()->signature(comp()->trMemory()));
-         traceMsg(comp(), "Returning...\n");
+         if (symRef->getSymbol()->isParm())
+            {
+            if (methodSymbol->sharesStackSlot(symRef))
+               {
+               sharesParm = true;
+               _sharedSymRefs.set(symRef->getReferenceNumber());
+               }
+            }
+         else if (symRef->getCPIndex() < methodSymbol->getFirstJitTempIndex())
+            {
+            containsAuto = true;
+            if (methodSymbol->sharesStackSlot(symRef))
+               _sharedSymRefs.set(symRef->getReferenceNumber());
+            }
          }
+      }
 
-      return 0;
+   TR_Array<List<TR::SymbolReference>> *pendingPushSymRefs = comp()->getMethodSymbol()->getPendingPushSymRefs();
```

Comment:
if we don't specifically need a TR_Array for perf reasons could we use an std::deque, std::list or similar with stack scoped memory to the same effect?